### DISCUSSION TITLE
Sending Controls/Events

### DISCUSSION THREAD

**Discussion by tjensen** [tjensen/fsuipc#14] [Ideas]
Controls (aka "Events") are sent via offset 0x3110:

![image](https://github.com/tjensen/fsuipc/assets/798166/5ddb8a4d-1b09-4cd0-b552-5b9266457702)

So, to send `8700` to control `84137`, you could do this:

```python
with FSUIPC() as f:
    f.write([(0x3110, "L", (8700 << 32) | 84137)])
```

Or, you could write a generic function to send controls:

```python
def send_control(f, control, parameter):
    data = parameter << 32 | control
    f.write([(0x3110, "L", data)])


with FSUIPC() as f:
  send_control(f, 84137, 8700)
```

**Comment by snizbatch** [comment]
tjensen, I'm trying to use FSUIPC to set 0x3110 to control the MCP Mach readout on the PMDG 777.  I have everything else working but this one is erratic and it is the only 4 byte FLT32 variable I'm trying to set.

mach_value = 830
base = 0x00011000  # integer 69632
OFFSET_MACH = 14503

f.write([(0x3110, "L", (mach_value << 32) | base + OFFSET_MACH)])

Do you have any ideas how I can set this successfully?  

Jim H.

**Reply by tjensen** [reply]
@snizbatch I know nothing about MCP Mach and PMDG 777 so this is just a guess...

If `mach_value` needs to be a 32-bit float, then I presume you need to encode it properly in the upper dword.

Try changing:
```python
mach_value = 830
```
to:
```python
import struct

mach_value = struct.unpack("I", struct.pack("f", 830))[0]
```

**Comment by snizbatch** [comment]
Tim,
   Thanks for the reply!  I did some playing around with struct pack and unpack before your comment to try to format the integer properly but ended up back at square one.  It actually turns out that the SDK shows the float value being set by a 2 digit integer but read as a 32 bit float.  I need to contact the publisher (PMDG) and find out why.  Thank you for your help in the meantime!

Jim Hunt

Sent from Mail for Windows
